lib/msf/core/payload/adapter/fetch.rb

```diff
@@ -0,0 +1,130 @@
+      comm = framework.sessions.get(srv_comm.to_i)
+      raise(RuntimeError, "Socket Server Comm (Session #{srv_comm}) does not exist") unless comm
+      raise(RuntimeError, "Socket Server Comm (Session #{srv_comm}) does not implement Rex::Socket::Comm") unless comm.is_a? ::Rex::Socket::Comm
+    when nil, ''
+      unless ip.nil?
+        comm = Rex::Socket::SwitchBoard.best_comm(ip)
+      end
+    else
+      raise(RuntimeError, "SocketServer Comm '#{srv_comm}' is invalid")
+    end
+
+    comm || ::Rex::Socket::Comm::Local
+  end
+
+  def srvhost
+    datastore['FETCH_SRVHOST']
+  end
+
+  def srvport
+    datastore['FETCH_SRVPORT']
+  end
+
+  def fetch_bindhost
+    datastore['FetchListenerBindAddress'].blank? ? srvhost : datastore['FetchListenerBindAddress']
+  end
+
+  def fetch_bindport
+    datastore['FetchListenerBindPort'].blank? ? srvport : datastore['FetchListenerBindPort']
+  end
+
+  # If no fetch URL is provided, we generate one based off the underlying payload data
+  # This is because if we use a randomly-generated URI, the URI generated by venom and
+  # Framework will not match.  This way, we can build a payload in venom and a listener
+  # in Framework, and if the underlying payload type/host/port are the same, the URI
+  # will be, too.
+  #
+  def default_srvuri
+    # If we're in framework, payload is in datastore; msfvenom has it in refname
+    payload_name = datastore['payload'] ||= refname
+    decoded_uri = payload_name.dup
+    # there may be no transport, so leave the connection string off if that's the case
+    if module_info['ConnectionType'].upcase == 'REVERSE'
+      decoded_uri << datastore['LHOST']
+      decoded_uri << datastore['LPORT'].to_s
+    end
+    if module_info['ConnectionType'].upcase == 'BIND'
+      decoded_uri << datastore['LHOST']
+      decoded_uri << datastore['RPORT'].to_s
+    end
+    Rex::Text.encode_base64url(decoded_uri)[0, 12]
+  end
+
+  def srvuri
+    return datastore['FETCH_URIPATH'] unless datastore['FETCH_URIPATH'].empty?
+    default_srvuri
+  end
+
+  def handle_connection(conn, opts = {})
+    opts[:arch] ||= module_info['AdaptedArch']
+    super
+  end
+
+  def compatible?(mod)
+    if mod.type == Msf::MODULE_PAYLOAD && (mod.class.const_defined?(:CachedSize) && mod.class::CachedSize != :dynamic) && (mod.class::CachedSize >= 120_000) # echo does not have an unlimited amount of space
+      return false
+    end
+    super
+  end
+  
+  def generate_payload_uuid(conf = {})
+    conf[:arch] ||= module_info['AdaptedArch']
+    conf[:platform] ||= module_info['AdaptedPlatform']
+    super
+  end
+
+  def check_srvhost
+    if Rex::Socket.is_ip_addr?(srvhost) && Rex::Socket.addr_atoi(srvhost) == 0
+      raise ArgumentError, 'You must set FETCH_SRVHOST to a routable IP'
+    end
+  end
+
+  def srvname
+    datastore['FetchServerName']
+  end
+
+  def generate_stage(opts = {})
+    opts[:arch] ||= module_info['AdaptedArch']
+    super
+  end
+
+
+  def remote_destination
```

Comment:
If we add quotation marks to the name, it makes it so we can handle spaces in the path and filename, but adds complexity to the escapes we need.